designs/2025-base-path-in-config-objects/README.md

```diff
@@ -0,0 +1,89 @@
+
+This RFC proposes allowing config objects to specify their own `basePath` property. When present, config object's `basePath` overrides config array's `basePath`, meaning that `files` and `ignores` patterns in that config object should be treated as relative to the config object's `basePath` instead of the config array's `basePath`.
+
+The new `basePath` property of config objects can be an absolute path, which is primarily intended for internal use by `eslint`, or a relative path, which is primarily intended for end users. A relative path is interpreted as relative to the config array's `basePath`.
+
+## Motivation
+
+Patterns passed through the `--ignore-pattern` CLI option / `ignorePatterns` API option should be treated as relative to the current working directory. When the current working directory must be the same or descendant of the base path directory, it is easy to transform the patterns by prepending the relative path from the base path directory to the current working directory, which is how it is [currently implemented](https://github.com/eslint/eslint/blob/03fb0bca2be41597fcea7c0e84456bbaf2e5acca/lib/config/config-loader.js#L568-L604).
+
+However, with the [new configuration file resolution](https://github.com/eslint/rfcs/tree/main/designs/2024-config-lookup-from-file) (currently available with the `unstable_config_lookup_from_file` flag; it will become the default behavior in ESLint v10), the current working directory no longer has to be a descendant of the base path directory, which makes pattern transformations very difficult, if not theoretically impossible. Instead of transforming patterns, we will introduce a new property that represents a directory to which the patterns are relative.
+
+Additional use cases may include user-specific scenarios, like manually merging configuration files from different directories, in which case specifying a `basePath` for imported patterns would be helpful to avoid transforming them manually.
+
+## Detailed Design
+
+Proof of concept:
+
+- `@eslint/config-array` changes: https://github.com/mdjermanovic/rewrite/pull/1
+- `eslint` changes: https://github.com/mdjermanovic/eslint/pull/6
+
+Most of the changes will be in the `@eslint/config-array` package.
+
+### Changes in `@eslint/config-array`
+
+- `baseSchema` will be updated to allow the new `basePath` property in config objects.
+- `filesAndIgnoresSchema` will be updated to validate that the new `basePath` property, if present, has a string value.
+- `normalizeConfigPatterns()` will be updated to normalize the `basePath` property value to an equivalent namespace-prefixed path, for consistency with the config array's `basePath` property value.
+- `META_FIELDS` will be updated to include `"basePath"`, in order to treat config objects with `basePath` + `ignores` as global ignores.
+- `get ignores()` will be updated to return an array of config objects instead of an array of ignore patterns.
```

Comment:
Presumably this is because we need to combine the `basePath` with the ignore patterns to get the correct files/directories to ignore? If so, can you add that detail here?

Reply:
Exactly, I have now added the explanation in f6d0f952a8aa85953380c2c7cdac68cbf209ee20.